source/slang/slang-ir-specialize-function-call.cpp

```diff
@@ -559,6 +561,21 @@ struct FunctionParameterSpecializationContext
             // the arguments at the new call site, and
             // don't add anything to the specialization key.
             //
+            // We should also add 2 more things such that our specialization
+            // can handle the corner cases that if the oldBase is a nonuniform
+            // resource and also the data type of oldIndex will be handled correctly.
+            // By doing so, we form an IRAttributedType to include both information
+            // and add it to the key of call info.
+
+            List<IRAttr*> irAttrs;
+            if (oldIndex->getOp() == kIROp_NonUniformResourceIndex)
```

Comment:
we probably also need to scan through any int casts to discover the nonuniform inst.
e.g. IntCast(IntCast(NonUniform(x))) is still a nonuniform.

Reply:
Is `cast` the only operator that doesn't change the uniformity?
How about others?
e.g. NonUniform(x)+1, NonUniform(x) * 2?

I'd like to handle them all.

Reply:
I think just handling cast right now would be good enough.
The general case should be handle with a proper data flow analysis, and I am fine if we scope that out of this PR.

Reply:
OK, push a new commit to cover this.